scripts/build-dist.sh

```diff
@@ -1,23 +1,40 @@
-#!/bin/sh
-
+#!/bin/bash
 set -ex
+# Builds the release tarball for Yarn.
+
+umask 0022 # Ensure permissions are correct (0755 for dirs, 0644 for files)
 
-npm run build
-npm pack
+# Workaround for https://github.com/yarnpkg/yarn/issues/2591
+case "$(uname -s)" in
+  *CYGWIN*|MSYS*|MINGW*)
+    dist_yarn=dist/bin/yarn.cmd
+    system_yarn=yarn.cmd
+    ;;
+  *)
+    dist_yarn=dist/bin/yarn
+    system_yarn=yarn
+    ;;
+esac
+
+rm -rf artifacts dist
 rm -rf dist
-mkdir dist
-mkdir -p artifacts
-mv yarn-*.tgz dist/pack.tgz
+mkdir artifacts
+mkdir dist{,/bin,/lib}
 
-cd dist
-umask 0022 # Ensure permissions are correct (0755 for dirs, 0644 for files)
-tar -xzf pack.tgz --strip 1
-rm -rf pack.tgz
-# Change this to "yarn install --production" once #1115 is fixed
-npm install --production
-../scripts/clean-node-modules.sh
-../scripts/set-installation-method.js $(readlink -f package.json) tar
-cd ..
+# Workaround for https://github.com/yarnpkg/yarn/issues/2591
+eval $system_yarn run build
+eval $system_yarn run build-bundle
+chmod +x artifacts/*.js
+
+cp package.json dist/
+cp LICENSE dist/
+cp artifacts/yarn-legacy-*.js dist/lib/yarn-cli.js
+cp bin/yarn-bundle-entry.js dist/bin/yarn.js
+cp bin/{yarn,yarnpkg,*.cmd} dist/bin/
+cp -r bin/node-gyp-bin dist/bin/
```

Comment:
Actually I wonder if Yarn should install node-gyp when needed (ie. run `yarn global add node-gyp`) rather than bundling it? I think bundling node-gyp would also require us to bundle all of its dependencies, so we'll end up with a giant node_modules directory bundled with Yarn again :(